src/main/java/edu/harvard/iq/dataverse/api/errorhandlers/WebApplicationExceptionHandler.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package edu.harvard.iq.dataverse.api.errorhandlers;
+
+import edu.harvard.iq.dataverse.api.util.JSONResponseBuilder;
+import edu.harvard.iq.dataverse.util.BundleUtil;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.swing.text.html.Option;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Catches all types of web application exceptions like NotFoundException, etc etc and handles them properly.
+ */
+@Provider
+public class WebApplicationExceptionHandler implements ExceptionMapper<WebApplicationException> {
+    
+    static final Logger logger = Logger.getLogger(WebApplicationExceptionHandler.class.getSimpleName());
+    
+    @Context
+    HttpServletRequest request;
+
+    @Override
+    public Response toResponse(WebApplicationException ex) {
+        
+        // If this is not a HTTP client or server error, just pass the response.
+        if (ex.getResponse().getStatus() < 400)
+            return ex.getResponse();
+        
+        // Otherwise, do stuff.
+        JSONResponseBuilder jrb = JSONResponseBuilder.error(ex.getResponse());
+        
+        // See also https://en.wikipedia.org/wiki/List_of_HTTP_status_codes for a list of status codes.
+        switch (ex.getResponse().getStatus()) {
+            // BadRequest
+            case 400:
+                if ( (ex.getMessage()+"").toLowerCase().startsWith("tabular data required")) {
+                    jrb.message(BundleUtil.getStringFromBundle("access.api.exception.metadata.not.available.for.nontabular.file"));
+                } else {
+                    jrb.message("Bad Request. The API request cannot be completed with the parameters supplied. Please check your code for typos, or consult our API guide at http://guides.dataverse.org.");
+                    jrb.request(request);
+                }
+                break;
+            // Forbidden
+            case 403:
+                jrb.message("Not authorized to access this object via this API endpoint. Please check your code for typos, or consult our API guide at http://guides.dataverse.org.");
```

Comment:
per my review comment of changing the AbstractApiBean.error method - would it make sense to move these status-specific messages to the JSON builder? That way they'd apply to calls from error() as well versus only happening on the exception route.

Reply:
Yes, that makes sense as far as I can foresee now.

But for the moment I would leave them there until it becomes clearer how the API response stuff will be done in the future.
(If you ask me, the errors should be handled completely by throwing JAX-RS or POJO exceptions, so there is a single place to run to and look at, avoiding to many places where stuff is wrapped in JSON sauce.)